learn_evm/arithmetic-checks.md

```diff
@@ -0,0 +1,717 @@
+function checkedMulInt64(int64 a, int64 b) public pure returns (int64 c) {
+    unchecked {
+        int256 uc = int256(a) * int256(b);
+
+        // If the product can not be represented with 64 bits,
+        // there is overflow.
+        if (overflowInt64(uc)) arithmeticError();
+
+        c = int64(uc);
+    }
+}
+```
+
+However, if the maximum value of a product exceeds 256 bits, then this method won't be effective.
+This happens, for instance, when working with int192. The product `type(int192).min * type(int192).min` requires 192 + 192 = 384 bits to be stored, which exceeds the maximum of 256 bits.
+Overflow occurs in 256 bits, causing a loss of information, and it won't be logical to check if the result fits into 192 bits.
+In this scenario, we can rely on the previous checks and, for example, attempt to reconstruct one of the multiplicands.
+
+```solidity
+function checkedMulInt192_1(int192 a, int192 b) public pure returns (int192 c) {
+    unchecked {
+        c = a * b;
+
+        if (a != 0 && b != c / a) arithmeticError();
+        if (a = -1 && b == type(int192).min) arithmeticError();
+    }
+}
+```
+
+We must consider the two special circumstances:
+
+1. When one of the multiplicands is zero (`a == 0`), the other multiplicand cannot be retrieved. However, this case never results in overflow.
+2. Even if the multiplication is correct in 256 bits, the calculation overflows when only examining the least-significant 192 bits if the first multiplicand is negative one (`a = -1`) and the other multiplicand is the minimum value.
+
+An example might help explain the second case.
+
+```solidity
+  0xffffffffffffffff800000000000000000000000000000000000000000000000 // type(int192).min
+* 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff // -1
+= 0x0000000000000000800000000000000000000000000000000000000000000000 // type(int192).min (when seen as a int192)
+```
+
+A method to address this issue is to always start by sign-extending or cleaning the result before attempting to reconstruct the other multiplicand.
+By doing so, it eliminates the need to check for the special condition.
+
+```solidity
+/// @notice version >= 0.8.17
+function checkedMulInt192_2(int192 a, int192 b) public pure returns (int192 c) {
+    unchecked {
+        bool overflow;
+
+        assembly {
+            // Extend the sign for int192 (24 = 23 + 1 bytes).
+            c := signextend(23, mul(a, b))
+
+            // Overflow, if `a != 0 && b != c / a`.
+            overflow := iszero(or(iszero(a), eq(b, sdiv(c, a))))
+        }
+
+        if (overflow) arithmeticError();
+    }
+}
+```
+
+## Conclusion
+
+In conclusion, we hope this article has served as an informative guide on signed integer arithmetic within the EVM and the two's complement system. We have seen the added complexity that is introduced when performing checked arithmetic on signed integers compared to unsigned ones and when dealing with sub-32 byte types.
```

Comment:
maybe we should use a bullet list here to highlight the lessons learned - I think it would make it more readable.